### PR TITLE
[primTorch] Add ref for `triplet_margin_loss`, improve `triplet_margin_with_distance_loss`

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #85614

### CODE REVIEW

File: torch/_refs/nn/functional/__init__.py [hunk 1]
```diff
@@ -457,6 +459,83 @@ def threshold(
...
+    ] = None,
+    margin: float = 1.0,
+    swap: bool = False,
+    reduction: str = "mean",
+) -> TensorLikeType:
+    _check_reduction_value(reduction)
+
+    # TODO: Regular PyTorch has this check for triplet_margin_loss but not for
+    # triplet_margin_with_distance_loss.  However, the docs for both suggest
+    # these must have the same shape.
```
Comment (lezcano) [inline torch/_refs/nn/functional/__init__.py:510]: Improved error checking is always desirable. Cool!

File: torch/_refs/nn/functional/__init__.py [hunk 2]
```diff
@@ -457,6 +459,83 @@ def threshold(
...
+    a_dim = anchor.dim()
+    p_dim = positive.dim()
+    n_dim = negative.dim()
+    check(
+        a_dim == p_dim and p_dim == n_dim,
+        lambda: f"All inputs should have same dimension but got {a_dim}D, {p_dim}D and {n_dim}D inputs.",
+    )
+
+    # TODO: Additional check not implemented in regular PyTorch (as per the
+    # docs).
```
Comment (lezcano) [inline torch/_refs/nn/functional/__init__.py:520]: No need to note all the things that are added really. Just note (perhaps with a gh comment?) things that may be controversial. All these seem very reasonable to me.

**Review by lezcano — APPROVED** [review]
Cool, thanks!

I still have the same question for @mruberry. Should we implement in PrimTorch things that are already implemented in Python in core, or what's the policy around that?

**Comment by mruberry** [comment]
> Cool, thanks!
> 
> I still have the same question for @mruberry. Should we implement in PrimTorch things that are already implemented in Python in core, or what's the policy around that?

Yes! But I would say the priority is for operations implemented in C++ today

File: torch/_refs/nn/functional/__init__.py [hunk 3]
```diff
@@ -457,6 +459,83 @@ def threshold(
     return torch.where(a <= threshold, value, a)
 
 
+# CompositeImplicitAutograd - don't register decomp
+# No elementwise type promotion here because it breaks a bunch of dtype tests.
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:463]: Let's elaborate on this -- does `torch.triplet_margin_loss` type promote?

File: torch/_refs/nn/functional/__init__.py [hunk 4]
```diff
@@ -457,6 +459,83 @@ def threshold(
...
+    margin: float = 1.0,
+    p: float = 2,
+    eps: float = 1e-6,
+    swap: bool = False,
+    size_average: Optional[bool] = None,
+    reduce: Optional[bool] = None,
+    reduction: str = "mean",
+) -> TensorLikeType:
+    if size_average is not None or reduce is not None:
+        # TODO: raise exception instead of converting value
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:477]: This should elaborate on this TODO -- is this a TODO in eager mode, too?

File: torch/_refs/nn/functional/__init__.py [hunk 5]
```diff
@@ -457,6 +459,83 @@ def threshold(
...
+    distance_function: Optional[
+        Callable[[TensorLikeType, TensorLikeType], TensorLikeType]
+    ] = None,
+    margin: float = 1.0,
+    swap: bool = False,
+    reduction: str = "mean",
+) -> TensorLikeType:
+    _check_reduction_value(reduction)
+
+    # TODO: Regular PyTorch has this check for triplet_margin_loss but not for
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:508]: Nice comment to help us reviewers, but it can be removed

File: torch/_refs/nn/functional/__init__.py [hunk 6]
```diff
@@ -457,6 +459,83 @@ def threshold(
...
+    # triplet_margin_with_distance_loss.  However, the docs for both suggest
+    # these must have the same shape.
+    a_dim = anchor.dim()
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:511]: Style nit: prefer using the `ndim` property (personally I would prefer we have a `rank` property, but oh well)

File: torch/_refs/nn/functional/__init__.py [hunk 7]
```diff
@@ -457,6 +459,83 @@ def threshold(
...
+    p_dim = positive.dim()
+    n_dim = negative.dim()
+    check(
+        a_dim == p_dim and p_dim == n_dim,
+        lambda: f"All inputs should have same dimension but got {a_dim}D, {p_dim}D and {n_dim}D inputs.",
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:516]: "All inputs" -> "The anchor, positive, and negative tensors"
"should have same dimension" -> "are expected to have the same number of dimensions"

Also add the parameter names so users know the rank of each tensor, otherwise the message just says that at least one tensor has a distinct rank, but not which one

File: torch/_refs/nn/functional/__init__.py [hunk 8]
```diff
@@ -457,6 +459,83 @@ def threshold(
...
+    # are not very clear on that, so not doing it to avoid breakage.  If this
+    # code is reused, make sure the behavior is compatible with
+    # triplet_margin_loss and triplet_margin_with_distance_loss.
+
+    if distance_function is None:
+        distance_function = torch.pairwise_distance
+
+    dist_pos = distance_function(anchor, positive)
+    dist_neg = distance_function(anchor, negative)
+    if swap:
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:532]: Add a comment for what `swap` does

File: torch/testing/_internal/common_methods_invocations.py [hunk 9]
```diff
@@ -17375,6 +17425,34 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
         torch_opinfo_name="clamp",
         supports_nvfuser=False,
     ),
+    PythonRefInfo(
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:17428]: So back to @lezcano's point about operators written in Python. Now that I've had a chance to review the PR let me elaborate on my previous comment and see if @lezcano agrees with this perspective.

The core idea of primTorch is that, ultimately, PyTorch programs can be represented using a smaller set of operators that more simply describe the program's semantics. So it's natural to write operations using those primitives and then operations using those operations and so on. 

PyTorch operations that are entirely defined in Python are effectively user-defined operations that live under the `torch` namespace. We shouldn't need to add explicit Python references for them because they are their own Python references. However, as we're seeing with many other PyTorch operators, there are often opportunities to improve these operations and/or their tests to be more consistent with the detailed attention we should be paying to Python references.

Because of that, I don't think we want to add these as Python references. Instead my question is: why not update the actual operator implementations with the changes made here? Does `_apply_loss_reduction` need to be updated to call `torch.sum` instead of `_refs.sum` directly? Are there other divergences? Is running the refs eagerly slower than the actual operation? Etc. As this PR points out there are fixes here for the original operations, and the test extension is nice, I just think we can drop addition of the `PythonRefInfo` and adding these operations to the `_refs` namespace. 

What are your thoughts, @nkaretnikov?

Comment (lezcano) [reply torch/testing/_internal/common_methods_invocations.py:17428]: As I see it, there's three options

1. [Mike's]: Not writing the ref at all, just have the OG Python version
2. [Extension of Mike's]: The ref should be an alias (`fn_alias = fn`) of the OG.
3. The main function should be in refs, and the function in the main namespace should be an alias to this function 

Between 1 and 2, I think that 2 is strictly better than 1. Also, everything that Mike mentioned applies to 2, so I'll just discuss 3.

So, the thing is that in `_refs` we do have some very nice tools developed to write operators in Python. `_apply_loss_reduction` and `utils.check` are two great examples of this.
As such, there may be some value in rewriting these operators as `_refs`. If we go down this route, then we surely will need to add matching `PythonRefInfo`, of course. If this is done an the ref includes improvements, these should also be reflected in the original version. To do this, we can simply write the original version as an alias to the ref (!!). This suggestion goes in line with the line of though of "PrimTorch will, at some point, become the main entry point to the PyTorch library".

An issue with this approach may be that we may run into circular imports, so I don't know how feasible is it.

Another issue related to the above is that we would start interweaving PrimTorch with core, which we may not want to do just yet.

The pros of "having all the implementations together in the same format" seems nice. For example, this function should probably need some broadcasting which it doesn't currently have. Now, the cons are also quite annoying, so I don't know. What do you guys think?

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:17428]: Good idea that we may want an alias, @lezcano -- as you say, why not?

For now I would suggesting going with approach (2). It's easy to switch to approach (3) later if we want, as that's just about what file the function is defined in. Having a single implementation of these Python-only operations seems like the thing.

I wouldn't worry too much about "interweaving primTorch with core." The primTorch project effectively has distinct parts, including the primitives themselves and a utility library for writing PyTorch operations in Python. I wouldn't want these operations to call prims directly at this time, but because the references are written as calling into the torch namespace I think components of the utility library are reusable without issue (like the functions you mention).

Typically with operations like this that are defined as composites PyTorch hasn't explicitly defined their broadcasting or type promotion behavior, and relied on the behavior inherited from the composition. This is imperfect since our type promotion behavior isn't composable, but it's been good enough and I don't think we should seek to change that standard with this PR. Also, the type promotion decorator we use with the primTorch Python references is probably too slow to use here.

Comment (nkaretnikov) [reply torch/testing/_internal/common_methods_invocations.py:17428]: > [Extension of Mike's]: The ref should be an alias (fn_alias = fn) of the OG.

I've tried this approach with `gaussian_nll_loss` from another PR that I have opened. It results in `maximum recursion depth exceeded` during testing (ref dispatches to orig impl which dispatches to ref ...). So I'd suggest to just skip adding references for pure Python ops unless we have a good reason to do it. We can always do it later since they all should be copy-pastable.

Personally, I'm not in favor of mixing regular PyTorch and primTorch just yet. We can do it later once the major issues are addressed. As of now, I feel like this mixing adds up to "maintenance cost" for no good reason.

So my vote is to go with this for now (while keeping a list of pure Python ops somewhere -- for later):
> [Mike's]: Not writing the ref at all, just have the OG Python version

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:17428]: Just updating the original Python version with these improvements and the test extension sounds great for now

File: torch/_refs/nn/functional/__init__.py (hunk 2 above)
Comment (nkaretnikov) [reply torch/_refs/nn/functional/__init__.py]: decided not to add the margin check because of the `max` in the formula

**Comment by nkaretnikov** [comment]
@mruberry PTAL

File: aten/src/ATen/native/Loss.cpp [hunk 10]
```diff
@@ -157,15 +157,17 @@ Tensor triplet_margin_loss(const Tensor& anchor, const Tensor& positive, const T
...
-      "D and ",
-      n_dim,
-      "D inputs.")
+      "The anchor, positive, and negative tensors are expected to have "
+      "the same number of dimensions, but got: anchor ", a_dim, "D, "
+      "positive ", p_dim, "D, and negative ", n_dim, "D inputs")
+
   auto dist_pos = at::pairwise_distance(anchor, positive, p, eps);
   auto dist_neg = at::pairwise_distance(anchor, negative, p, eps);
+  // The distance swap is described in the paper "Learning shallow
```
Comment (mruberry) [inline aten/src/ATen/native/Loss.cpp:166]: Nice comment

File: torch/_refs/nn/functional/__init__.py [hunk 11]
```diff
@@ -501,6 +504,83 @@ def threshold(
...
+        margin=margin,
+        swap=swap,
+        reduction=reduction,
+    )
+
+
+# TODO: Pure Python impl - don't register decomp and don't add a ref
+# Defined as a helper here because triplet_margin_loss can be nicely implemented
+# with it.
+def _triplet_margin_with_distance_loss(
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:541]: This is going to need a comment for why the Python reference for triplet margin loss can't just call `nn.functional.triplet_margin_with_distance_loss` directly

**Comment by nkaretnikov** [comment]
@mruberry PTAL

**Review by mruberry — APPROVED** [review]
Cool!

A nice follow-up would be adding reference inputs for triplet margin loss. The current sample inputs are look thorough but small
